Snapper execution module and state for Salt

meaksh opened the pull request:
### What does this PR do?

This PR includes the Snapper execution module for Salt. It allows to manage Snapper using the DBus interface and also implements `snapper.run` in order to wrap a `__salt__` function and create pre/post snapshots before and after the function is called. Then with `snapper.diff_jid` and `snapper.undo_jid` functions you can show the changes made by a `jid` and also undo these changes.

Available functions:

```
# Basic methods
snapper.list_snapshots
snapper.list_configs
snapper.get_config
snapper.set_config
snapper.get_snapshot
snapper.changed_files
snapper.create_snapshot
snapper.status
snapper.diff
snapper.undo

#  Call a `__salt__` function and creates a pre/post snapshots before and after execute it.
snapper.run  

# Show diff for a Salt `jid` called with snapper.run
snapper.diff_jid

# Undo the changes applied by a Salt `jid` called with snapper.run
snapper.undo_jid
```

This PR also includes a snapper state that allows to manage configuration drifts based on custom `baseline`  snapshots:

```
my_baseline:
  snapper.baseline_snapshot:
    - number: 20
    - ignore:
      - /var/log
      - /var/cache
```

Uses and examples: [Managing configuration drift with Salt and Snapper](https://duncan.mac-vicar.com/2016/06/09/config-drift-salt-snapper.html)
### Tests written?

Yes

/cc @dmacvicar @isbm @dincamihai

meaksh committed:
snapper execution module

meaksh committed:
snapper state module

meaksh committed:
snapper module unit tests

meaksh committed:
some pylint fixes

meaksh committed:
more unit tests

meaksh committed:
Fix for snapper.diff when files are created or deleted

meaksh committed:
fix diff unit test while creating text file

meaksh committed:
passing *args and **kwargs to function when snapper.run

meaksh committed:
unit test for snapper.diff with binary file

meaksh committed:
load snapper state only if snapper module is loaded

meaksh committed:
Fix for _get_jid_snapshots if snapshots doesn't exist

meaksh committed:
pylint fixes

meaksh committed:
pylint: some fixes

dmacvicar commented:
Great work @meaksh getting this ready.

Something that could be an addition before of after the merge is the feature of additionally allowing to specify the baseline snapshot based on a tag that can be moved (in addition to absolute snapshot number).

This could be done either very low level, eg. allow to specify a userdata to search.

```yaml
snapper.baseline_snapshot:
    - tag:
        baseline
```

or simpler, we add a `snapper.tag_baseline` that sets the right userdata key/value that the state will search. Any ideas on this from others?

meaksh committed:
Variable renaming. Pylint fixes

meaksh committed:
Fix in inline comments

meaksh committed:
Fix for pylint: W1699

cachedout commented on salt/modules/snapper.py:
```diff
@@ -0,0 +1,632 @@
+
+if HAS_DBUS:
+    bus = dbus.SystemBus()  # pylint: disable=invalid-name
+    snapper = dbus.Interface(bus.get_object('org.opensuse.Snapper',  # pylint: disable=invalid-name
```
You mention [here](https://github.com/saltstack/salt/pull/34599/files#diff-f944d9c406d81a5a49e5491953920273R9) that Snapper is available in most distros but this path looks SUSE-specific. We need to either explicitly restrict use of this module to SUSE or normalize this path for other distros.

dmacvicar replied:
@cachedout That path is the snapper dbus interface. It will be the same in every distro, just like firewallD dbus path is `org.fedoraproject.FirewallD1` in every distro.

cachedout replied:
Cool. I didn't know. That's excellent!

dmacvicar replied:
The snapper cli tools is just a thin client for the daemon dbus interface. The module also calls the dbus interface. The only operation we do with the command line tool is `undo` because the dbus interface for security reasons can't modify the system.

meaksh replied:
Thank you for the clarification here @dmacvicar . I was not sure if dbus path is preserved on every distro.

cachedout commented on salt/modules/snapper.py:
```diff
@@ -0,0 +1,632 @@
+def _snapshot_to_data(snapshot):
+    '''
+    Returns snapshot data from a D-Bus response
+    '''
```
It would be nice to describe the structure of a `snapshot` here. Looks like a list?

cachedout commented on salt/modules/snapper.py:
```diff
@@ -0,0 +1,632 @@
+    if not userdata:
+        userdata = {}
+
+    jid = kwargs.get('__pub_jid', None)
```
`None` is the default here, FWIW.

meaksh committed:
some fixes and comments improvement

cachedout commented:
We need to gate this test to systems where the necessary requirements are installed: https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/3404/testReport/junit/unittest.loader/ModuleImportFailure/snapper_test/

meaksh committed:
Prevent module failing if Snapper does not exist in D-Bus

meaksh committed:
Added function for baseline creation

meaksh committed:
Allow tag reference for baseline_snapshot state

meaksh commented:
Added additional support for baseline state definition using `tag` instead of `number` as @dmacvicar suggested.

Also added a fix to prevent the failure of the module when running in a machine without Snapper. This fixes the failing unit test pointed by @cachedout.

cachedout commented:
Go Go Jenkins!

cachedout commented:
Merged! Thanks, @meaksh !
